### DISCUSSION TITLE
Resume a stopped test session

### DISCUSSION THREAD

**Discussion by AndrewFalanga** [pytest-dev/pytest#8673] [Q&A] [answered]
I have been looking into this for a few hours now.  I haven't found much.

What I need to resume a session which was stopped, probably due to loss of power.  Is this possible?  I have found very little via many Google searches and time on Stackoverflow.  The closes I've found thus far is this http://lists.idyll.org/pipermail/testing-in-python/2018-January/007157.html.  At that time, the OP was told that "support for this didn't yet exist."  It's 3 years later, has this changed?

If it hasn't changed, what sort of things would be important if writing our own set of plugins to support such a feature?

Thanks,
Andrew

**Comment by nicoddemus** [comment, marked as answer]
Hi @AndrewFalanga,

Currently there's no native support for that, but the OP in the message comments that he managed to write a simple plugin that does this as he intends. 

I suggest to follow the same idea, should be relatively simple, here's a quick outline:

1. Implement `pytest_runtest_protocol` as a hookwrapper: when the wrapper "ends", it means the test ran to completion, so record the test node id in a specific file on disk.
2. When the test session finishes (implement `pytest_sessionfinish`), delete that file from disk to indicate a complete run.
3. When running for the second time, implement `pytest_collection_modifyitems`, which then should see if the file exists: if it does, it should remove the node ids from the collection that are in the file, because the file's existence indicates a previous run terminated abruptly, so those items have already executed and we should resume from that point.

Hope this helps!

**Reply by AndrewFalanga** [reply]
Thank @nicoddemus for the pointers on implementing the plugins which will be required to do these actions.  This will greatly reduce the time for implementation.

**Reply by AndrewFalanga** [reply]
@nicoddemus I've been working on this for a bit now and I have a question about properly deselecting these things.  My current process is to simply `<list>.pop()` the items which have the nodeid that I write to the file.  This _does work_ however it doesn't seem that it's the correct way to remove them.  How do I modify what is selected and deselected in a correct way?

**Reply by graingert** [reply]
have a look at the stepwise plugin https://github.com/pytest-dev/pytest/blob/bb9683deea7ca8aee482973d62202071ea9420e8/src/_pytest/stepwise.py#L65

**Reply by AndrewFalanga** [reply]
@graingert thank you.  That is very intriguing.

**Reply by graingert** [reply]
I think writing the test state (collected/started/passed/failed) to sqlite would probably be the safest

this way sqlite can handle terminating during a run safely without corrupting the file


I'd love something like `--stepwise` that works with pytest-xdist and pytest-randomly and it sounds like that's what you're writing?
